src/unix/solarish/compat.rs

```diff
@@ -169,3 +170,33 @@ pub unsafe fn forkpty(
 
     0
 }
+
+pub unsafe fn getpwent_r(
+    pwd: *mut passwd,
+    buf: *mut ::c_char,
+    buflen: ::size_t,
+    result: *mut *mut passwd,
+) -> ::c_int {
+    let res = native_getpwent_r(pwd, buf, buflen.try_into().unwrap());
+    if res.is_null() {
+        -1
+    } else {
+        *result = res;
+        0
+    }
+}
```

Comment:
Hey, sorry about this, but I was reading the manual page again today and I think I missed an edge case in my original suggestion.  Critically, when there are no more entries, for glibc:

```
RETURN VALUE
       On success, these functions return 0 and *pwbufp is a  pointer  to  the
       struct  passwd.   On  error,  these functions return an error value and
       *pwbufp is NULL.

ERRORS
       ENOENT No more entries.
```

Contrast this with the manual for the illumos version:

```
RETURN VALUES
...
       Applications wishing to check for error situations should set errno to
       0 before calling getpwnam(), getpwnam_r(), getpwuid(), getpwuid_r(),
       getpwent(), getpwent_r(), fgetpwent(), and fgetpwent_r(). If these non-
       reentrant functions return a null pointer and errno is non-zero, an
       error occurred.
```

So I think we need to clear `errno` ourselves here, and return `ENOENT` explicitly; i.e., something like:

```rust
*___errno() = 0;
let res = native_getpwent_r(pwd, buf, buflen.try_into().unwrap());
if res.is_null() {
    if *___errno() == 0 {
        *___errno() = ENOENT;
    }
    -1
} else {
    *result = res;
    0
}

Reply:
Yes, glibc returns 0 and sets `errno` to `ENOENT` when there are no more entries. But different UNIX (or UNIX-like) OSes actually have different behaviors in this case. For example:

* freebsd: [getpwent_r man page](https://www.freebsd.org/cgi/man.cgi?query=getpwent&sektion=3)

   On success: return 0, `result` will be non-NULL
   On error: return the error number, `result` will be NULL
   End: return 0, `result` will be NULL.

* dragonfly: [getpwent_r man page](https://leaf.dragonflybsd.org/cgi/web-man?command=getpwent_r&section=ANY)

  On success: return 0 and `result` will be non-NULL
  On error: return the error number, `result` will be NULL
  End: return 0, `result` will be NULL
  > The behavior on freebsd and dragonfly are the same.

* netbsd: [getpwent_r man page](https://man.netbsd.org/getpwent_r.3)

  On success: return 0, `result` will be non-NULL,
  On error: return a non-zero value (`errno` is set), `result` will be NULL
  End: return 0, `result` will be NULL

* Linux: [getpwent_r man page](https://man7.org/linux/man-pages/man3/getpwent_r.3.html)

  On success: return 0, `result` will be non-NULL
  On error: return the error number (**not ENOENT**), `result` will be NULL
  End: return `ENOENT` and `result` will be NULL.

Are we going to adopt the glibc one?

Reply:
If we follow the glibc one, then the impl will be something like:

```rust
pub unsafe fn getpwent_r(
    pwd: *mut passwd,
    buf: *mut ::c_char,
    buflen: ::size_t,
    result: *mut *mut passwd,
) -> ::c_int {
    * ::___errno() = 0;
    let res = native_getpwent_r(pwd, buf, buflen.min(::c_int::MAX as ::size_t) as ::c_int);
    if res.is_null() {
        *result = core::ptr::null_mut();
        if * ::___errno() == 0 {
            return ENOENT;
        } else {
            return * ::___errno();
        }
    } else {
        *result = res;
        0
    }
}
```

Reply:
Oh bother.  Thank you for comparing the BSDs, which I had not done.  That is somewhat unfortunate!  The FreeBSD behaviour seems to be the same as what the standardised `getpwnam_r()` and `getgrnam_r()` offer, which probably makes sense here.  Clearly this is always going to be an ifdef soup for people.

I think probably we should do what FreeBSD is doing here because it's so much like the existing standard, so, something like:

```rust
let olderror = *___errno();
*___errno() = 0;
*result = native_getpwent_r(pwd, buf, buflen);
let reterr = if (*result).is_null() {
    *___errno()
} else {
    0
};
*___errno() = olderror;
reterr
```

That way, we can use `errno` ourselves but not stomp on it, and presumably we can just be included in whatever conditional compilation in `nix`, et al, is already in-place for FreeBSD.  This is pretty much what our POSIX-compliant `getgrnam_r()` wrapper does in illumos today; i.e.,

```c
/*
 * POSIX.1c standard version of the function getgrnam_r.
 * User gets it via static getgrnam_r from the header file.
 */
int
__posix_getgrnam_r(const char *name, struct group *grp, char *buffer,
    size_t bufsize, struct group **result)
{
        int nerrno = 0;
        int oerrno = errno;

        if ((*result = getgrnam_r(name, grp, buffer, (uintptr_t)bufsize))
            == NULL) {
                        nerrno = errno;
        }
        errno = oerrno;
        return (nerrno);
}
```

Reply:
Done.